osbuild/buildroot.py

```diff
@@ -14,115 +19,169 @@
-        self.mounts = []
-        self.libdir = libdir
-        self.runner = runner
-
-        self.mount_root(root)
-        self.mount_var()
-
-    def mount_root(self, root):
-        for p in ["boot", "usr", "bin", "sbin", "lib", "lib64"]:
-            source = os.path.join(root, p)
-            target = os.path.join(self.root, p)
-            if not os.path.isdir(source) or os.path.islink(source):
-                continue # only bind-mount real dirs
-            os.mkdir(target)
-            try:
-                subprocess.run(["mount", "-o", "bind,ro", source, target], check=True)
-            except subprocess.CalledProcessError:
-                self.unmount()
-                raise
-            self.mounts.append(target)
-
-        if platform.machine() == "s390x" or platform.machine() == "ppc64le":
-            # work around a combination of systemd not creating the link from
-            # /lib64 -> /usr/lib64 (see systemd issue #14311) and the dynamic
-            # linker is being set to (/lib/ld64.so.1 -> /lib64/ld64.so.1)
-            # on s390x or /lib64/ld64.so.2 on ppc64le
-            # Therefore we manually create the link before calling nspawn
-            os.symlink("/usr/lib64", f"{self.root}/lib64")
-
-    def mount_var(self):
-        target = os.path.join(self.root, "var")
-        os.mkdir(target)
-        try:
-            subprocess.run(["mount", "-o", "bind", self.var, target], check=True)
-        except subprocess.CalledProcessError:
-            self.unmount()
-            raise
-        self.mounts.append(target)
-
-    def unmount(self):
-        for path in self.mounts:
-            subprocess.run(["umount", "--lazy", path], check=True)
-            os.rmdir(path)
-        self.mounts = []
-        if self.root:
-            shutil.rmtree(self.root)
-            self.root = None
-        if self.api:
-            shutil.rmtree(self.api)
-            self.api = None
-        if self.var:
-            shutil.rmtree(self.var)
-            self.var = None
+        self._rootdir = root
+        self._rundir = path
+        self._vardir = var
+        self._libdir = libdir
+        self._runner = runner
+
+    @staticmethod
+    def _mknod(path, name, mode, major, minor):
+        os.mknod(os.path.join(path, name),
+                 mode=(stat.S_IMODE(mode) | stat.S_IFCHR),
+                 device=os.makedev(major, minor))
+
+    def __enter__(self):
+        self._exitstack = contextlib.ExitStack()
+        with self._exitstack:
+            # We create almost everything directly in the container as temporary
+            # directories and mounts. However, for some things we need external
+            # setup. For these, we create temporary directories which are then
+            # bind-mounted into the container.
+            #
+            # For now, this includes:
+            #
+            #   * We create an API directory where the caller can place sockets
```

Comment:
Can we use fd passing instead of this now that we are no longer using nspawn? In that case, would we be able to disable network access? In a follow-up commit, of course.

Reply:
I think so, yes.